pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,460 @@
+
+        self.total_shots_used = 0
+        """int: total number of shots used across all iterations"""
+
+        # total number of iterations
+        self.k = 0
+        # Number of shots per parameter
+        self.s = None
+        # maximum number of shots required to evaluate across all parameters
+        self.max_shots = None
+
+        # Running average of the parameter gradients
+        self.chi = None
+        # Running average of the variance of the parameter gradients
+        self.xi = None
+
+        super().__init__(stepsize=stepsize)
+
+    @staticmethod
+    def weighted_random_sampling(qnodes, coeffs, shots, argnums, *args, **kwargs):
+        """Returns an array containing length ``shots`` single-shot estimates
+        of the Hamiltonian gradient. The shots are distributed randomly over
+        the terms in the Hamiltonian, as per a multinomial distribution.
+
+        Args:
+            qnodes (Sequence[.tape.QNode]): Sequence of QNodes, each one when evaluated
+                returning the corresponding expectation value of a term in the Hamiltonian.
+            coeffs (Sequence[float]): Sequences of coefficients corresponding to
+                each term in the Hamiltonian.
+            shots (int): The number of shots used to estimate the Hamiltonian expectation
+                value. These shots are distributed over the terms in the Hamiltonian,
+                as per a Multinomial distribution.
+            argnums (Sequence[int]): the QNode argument indices which are trainable
+            *args: Arguments to the QNodes
+            **kwargs: Keyword arguments to the QNodes
+
+        Returns:
+            array[float]: the single-shot gradients of the Hamiltonian expectation value
+        """
+
+        # determine the shot probability per term
+        prob_shots = np.abs(coeffs) / np.sum(np.abs(coeffs))
+
+        # construct the multinomial distribution, and sample
+        # from it to determine how many shots to apply per term
+        si = multinomial(n=shots, p=prob_shots)
+        shots_per_term = si.rvs()[0]
+
+        grads = []
+
+        for h, c, p, s in zip(qnodes, coeffs, prob_shots, shots_per_term):
+
+            # if the number of shots is 0, do nothing
+            if s == 0:
+                continue
+
+            # set the QNode device shots
+            h.device.shots = [(1, s)]
+
+            jacs = []
+            for i in argnums:
+                j = qml.jacobian(h, argnum=i)(*args, **kwargs)
+
+                if s == 1:
+                    j = np.expand_dims(j, 0)
+
+                # Divide each term by the probability per shot. This is
+                # because we are sampling one at a time.
+                jacs.append(c * j / p)
+
+            grads.append(jacs)
+
+        return [np.concatenate(i) for i in zip(*grads)]
+
+    @staticmethod
+    def check_device(dev):
+        r"""Verifies that the device used by the objective function is non-analytic.

```

Comment:
Will this be changing with the shots refractor?  All we need is the ability to set the shots in a device, even if it initializes as `analytic=True`.  

For example, we should be able to pass in `dev = qml.device('default.qubit', wires=n)`.  Though it defaults to analytic, we can change it to shots mode.

Reply:
Yes it will 🙏 

Which will be really nice, and actually simplify the logic here a lot; this entire method can be deleted.